### PR TITLE
Handle case when executable is not executable

### PR DESCRIPTION
**Pull request by WillKoehrsen**
- Changed error message if executable is not executable

Closes:[1159](https://github.com/pre-commit/pre-commit/issues/1159)

### CODE REVIEW

File: pre_commit/parse_shebang.py [hunk 1]
```diff
@@ -52,7 +52,7 @@ def _error(msg):
             _error('not found')
         return exe
     elif not os.access(orig, os.X_OK):
-        _error('not found')
+        _error('not executable')
```
Comment (asottile) [inline pre_commit/parse_shebang.py]: there should be a new case here, one for "does not exist" and one for "not executable"

the test you modified shows that this change is incorrect in fact

also you'd use `_error('is not executable')`

Comment (WillKoehrsen) [reply pre_commit/parse_shebang.py]: Thanks for the review. So the check for existence would be:
```
elif not os.access(orig, os.F_OK)
```
and the check for executability would be:
```
elif not os.access(orig.os.F_OK)
```
Using the following mode codes:
```
os.F_OK: Tests existence of the path.
os.R_OK: Tests readability of the path.
os.W_OK: Tests writability of the path.
os.X_OK: Checks if path can be executed.
```

Comment (asottile) [reply pre_commit/parse_shebang.py]: `os.path.isfile` is the usual way to check for a file

File: tests/parse_shebang_test.py [hunk 2]
```diff
@@ -88,7 +88,21 @@ def test_normexe_does_not_exist():
 def test_normexe_does_not_exist_sep():
     with pytest.raises(OSError) as excinfo:
         parse_shebang.normexe('./i-dont-exist-lol')
-    assert excinfo.value.args == ('Executable `./i-dont-exist-lol` not found',)
+    assert excinfo.value.args == (
+        'Executable `./i-dont-exist-lol` not found',
+    )
```
Comment (asottile) [inline tests/parse_shebang_test.py]: this test shouldn't change -- can you revert the changes there?

File: tests/parse_shebang_test.py [hunk 3]
```diff
@@ -88,7 +88,21 @@ def test_normexe_does_not_exist():
...
+
+
+def test_normexe_not_executable():
+    exe_path = os.path.abspath(
+        write_executable('/usr/bin/env sh', filename='run.myext'),
+    )
```
Comment (asottile) [inline tests/parse_shebang_test.py]: this pollutes the working directory, can you use `tmpdir` to do this?

File: tests/parse_shebang_test.py [hunk 4]
```diff
@@ -88,7 +88,21 @@ def test_normexe_does_not_exist():
...
+def test_normexe_not_executable(in_tmpdir):
+    exe_path = os.path.abspath(
+        write_executable('/usr/bin/env sh', filename='run.myext'),
+    )
+    os.chmod(exe_path, 444)
+    with pytest.raises(OSError) as excinfo:
+        parse_shebang.normexe(exe_path)
+    assert excinfo.value.args == (
+        'Executable `{}` is not executable'.format(exe_path),
+    )
```
Comment (asottile) [inline tests/parse_shebang_test.py]: I suspect you might need to skip this test on windows given the way executables work there

you can use `# pragma: windows no cover` this test (and probably the code branch as well if it isn't going to be hit)

File: pre_commit/parse_shebang.py (hunk 1 above)
Comment (WillKoehrsen) [reply pre_commit/parse_shebang.py]: `os.path.isfile` will also catch a directory which is covered in `elif os.path.isdir(orig)`.

Comment (asottile) [reply pre_commit/parse_shebang.py]: you may need to reorder the checks then

File: tests/parse_shebang_test.py [hunk 5]
```diff
@@ -91,6 +91,18 @@ def test_normexe_does_not_exist_sep():
     assert excinfo.value.args == ('Executable `./i-dont-exist-lol` not found',)
 
 
+def test_normexe_not_executable(in_tmpdir):
+    exe_path = os.path.abspath(
+        write_executable('/usr/bin/env sh', filename='run.myext'),
+    )
+    os.chmod(exe_path, 444)
```
Comment (asottile) [inline tests/parse_shebang_test.py]: simpler:

```python
in_tmpdir.join('exe').ensure()
with pytest.raises(OSError) as excinfo:
    parse_shebang.normexe('./exe')
assert excinfo.value.args == ('Executable `./exe` is not executable',)
```

Comment (WillKoehrsen) [reply tests/parse_shebang_test.py]: Hmm, I get `AttributeError: 'str' object has no attribute 'ensure'` when I try to run that test.

Comment (asottile) [reply tests/parse_shebang_test.py]: ah right, this is before I switched to using the `tmpdir` fixture more -- so instead:

```python
def test_...(tmpdir):
    tmpdir.join('exe').ensure()
    with tmpdir.as_cwd(), pytest.raises(OSError) as excinfo:
        parse_shebang.normexe('./exe')
    assert excinfo.value.args == ('Executable `./exe` is not executable',)
```

File: tests/parse_shebang_test.py [hunk 6]
```diff
@@ -91,6 +91,13 @@ def test_normexe_does_not_exist_sep():
     assert excinfo.value.args == ('Executable `./i-dont-exist-lol` not found',)
 
 
+def test_normexe_not_executable(tmpdir):
```
Comment (asottile) [inline tests/parse_shebang_test.py]: you'll need `@pytest.mark.xfail(os.name == 'nt', reason='posix only')  # pragma: windows no cover`

sorry forgot about the test-skipping part

File: tests/parse_shebang_test.py [hunk 7]
```diff
@@ -91,6 +91,16 @@ def test_normexe_does_not_exist_sep():
     assert excinfo.value.args == ('Executable `./i-dont-exist-lol` not found',)
 
 
+@pytest.mark.xfail(
+    os.name == 'nt', reason='posix only',
+)  # pragma: windows no cover
```
Comment (asottile) [inline tests/parse_shebang_test.py]: here's another way to write this, nbd though:

```diff
diff --git a/tests/parse_shebang_test.py b/tests/parse_shebang_test.py
index 3c48a11..fe1cdcd 100644
--- a/tests/parse_shebang_test.py
+++ b/tests/parse_shebang_test.py
@@ -91,10 +91,8 @@ def test_normexe_does_not_exist_sep():
     assert excinfo.value.args == ('Executable `./i-dont-exist-lol` not found',)
 
 
-@pytest.mark.xfail(
-    os.name == 'nt', reason='posix only',
-)  # pragma: windows no cover
-def test_normexe_not_executable(tmpdir):
+@pytest.mark.xfail(os.name == 'nt', reason='posix only')
+def test_normexe_not_executable(tmpdir):  # pragma: windows no cover
     tmpdir.join('exe').ensure()
     with tmpdir.as_cwd(), pytest.raises(OSError) as excinfo:
         parse_shebang.normexe('./exe')
```

**Comment by WillKoehrsen** [comment]
@asottile Thanks for coaching me through this one! 👍

**Comment by asottile** [comment]
> @asottile Thanks for coaching me through this one! +1

happy to help! thanks for the contribution!
